src/gba_cell.rs

```diff
@@ -1,66 +1,79 @@
+/// size. Generally that means it should be one of:
 ///
-/// * LLVM thinks that ARMv4T only supports atomic operations via special atomic
-///   support library functions. This is true for the "complex" atomic ops like
-///   "fetch-add", but for individual load or store ops this is overkill.
-/// * If you directly write an Acquire/load, Release/store, or a Relaxed op with
-///   an associated `compiler_fence`, then LLVM does generate correct code.
-///   However, it will have very sub-optimal performance. LLVM will generate
-///   calls to the mythical atomic support library, when it should just directly
-///   use an `ldr` or `str` instruction.
-/// * In response to this LLVM nonsense, the `GbaCell` type just uses inline
-///   assembly to perform all accesses to the contained data.
-/// * When LLVM sees inline assembly, it is forced to defensively act as if the
-///   inline assembly might have done *anything* legally possible using the
-///   pointer and value provided to the inline assembly. This includes that the
-///   inline assembly *might* call the atomic support library to access the
-///   pointer's data using an atomic load or store. So LLVM has to treat the
-///   inline assembly as an atomic sync point.
-/// * However, inside the inline asm block we actually just use the single load
-///   or store op that we wanted.
+/// * an 8, 16, or 32 bit integer
+/// * a function pointer
+/// * a data pointer to a sized type
+/// * an optional non-null pointer (to function or sized data)
+/// * a `repr(transparent)` newtype over one of the above
+pub unsafe trait GbaCellSafe: Copy {}
+// Note(Lokathor): The list here is not exhaustive, it's just all the stuff I
+// thought of at the time. Add more as necessary.
+unsafe impl GbaCellSafe for u8 {}
+unsafe impl GbaCellSafe for u16 {}
+unsafe impl GbaCellSafe for u32 {}
+unsafe impl GbaCellSafe for i8 {}
+unsafe impl GbaCellSafe for i16 {}
+unsafe impl GbaCellSafe for i32 {}
+unsafe impl GbaCellSafe for NonZeroI16 {}
+unsafe impl GbaCellSafe for NonZeroI32 {}
+unsafe impl GbaCellSafe for NonZeroI8 {}
+unsafe impl GbaCellSafe for NonZeroU16 {}
+unsafe impl GbaCellSafe for NonZeroU32 {}
+unsafe impl GbaCellSafe for NonZeroU8 {}
+unsafe impl GbaCellSafe for Option<NonZeroI16> {}
+unsafe impl GbaCellSafe for Option<NonZeroI32> {}
+unsafe impl GbaCellSafe for Option<NonZeroI8> {}
+unsafe impl GbaCellSafe for Option<NonZeroU16> {}
+unsafe impl GbaCellSafe for Option<NonZeroU32> {}
+unsafe impl GbaCellSafe for Option<NonZeroU8> {}
+unsafe impl GbaCellSafe for char {}
+unsafe impl GbaCellSafe for bool {}
+unsafe impl GbaCellSafe for Color {}
+unsafe impl GbaCellSafe for KeyInput {}
+unsafe impl<T> GbaCellSafe for *const T {}
+unsafe impl<T> GbaCellSafe for *mut T {}
+unsafe impl<T> GbaCellSafe for NonNull<T> {}
+unsafe impl<T> GbaCellSafe for Option<NonNull<T>> {}
+unsafe impl GbaCellSafe for Option<unsafe extern "C" fn(u16)> {}
+unsafe impl GbaCellSafe for Option<unsafe extern "C" fn(IrqBits)> {}
```

Comment:
With nightly features, it's possible to write this impl but generic over the function arguments and return type.